### PR TITLE
added missing space in 'configopts' in ParaView 5.8.0 easyconfigs using 2020a toolchain

### PR DESCRIPTION
**Pull request by mboisson**
There's a missing space at the end of one line of `configopts`, which means that CMake options get ignored.

### CODE REVIEW

**Comment by mboisson** [comment]
@boegel, you missed something there ;)

**Comment by boegel** [comment]
@mboisson Ouch...

Any idea if we can catch this by enhancing the sanity check?

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in this PR)
generoso-x-5 - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2667 v3 @ 3.20GHz (haswell), Python 3.6.8
See https://gist.github.com/536d329c13f65909153520762d4881a8 for a full test report.

**Comment by smoors** [comment]
@boegel one solution would be to allow `buildopts` and friends be defined a list of strings, which get joined by EB with a space

**Comment by smoors** [comment]
Test report by @smoors
**FAILED**
Build succeeded for 3 out of 4 (2 easyconfigs in this PR)
node381.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz (skylake_avx512), Python 2.7.5
See https://gist.github.com/7b760e6fd4c597de9b5a98c990224900 for a full test report.

**Comment by mboisson** [comment]
Mmmm, so, are those failing tests just a symptoms of the corrected recipe not working, and the incorrect recipe skipping something ?

**Comment by smoors** [comment]
this is the error for the intel version (foss builds fine). looks indeed like an issue with OSMesa linking, which was skipped previously.
```
[ 65%] Building CXX object CMakeFiles/vtkFiltersPointsCS.dir/CMakeFiles/vtkFiltersPointsCS/vtkLinearKernelClientServer.cxx.o
/apps/brussel/CO7/skylake/software/impi/2019.7.217-iccifort-2020.1.217/intel64/bin/mpiicpc   -I/tmp/3254107.master01.hydra.brussel.vsc/tmp/vsc10009/build/ParaView/5.8.0/intel-2020a-Python-3.8.2-mpi/easybui
ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaDestroyContext'
ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaMakeCurrent'
ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaGetCurrentContext'
ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaCreateContext'
ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaGetProcAddress'
make[2]: *** [bin/vtkProbeOpenGLVersion-pv5.8] Error 1
make[2]: Leaving directory `/tmp/3254107.master01.hydra.brussel.vsc/tmp/vsc10009/build/ParaView/5.8.0/intel-2020a-Python-3.8.2-mpi/easybuild_obj'
make[1]: *** [VTK/Rendering/OpenGL2/CMakeFiles/vtkProbeOpenGLVersion.dir/all] Error 2

```

**Comment by boegel** [comment]
> @boegel one solution would be to allow `buildopts` and friends be defined a list of strings, which get joined by EB with a space

`buildopts` can already be a list of string values, but that's used for indicating that the configure-build-test-install cycle myst be run multiple times during a single installation...

**Comment by boegel** [comment]
> @boegel one solution would be to allow `buildopts` and friends be defined a list of strings, which get joined by EB with a space

`buildopts` can already be a list of string values, but that's used for indicating that the configure-build-test-install cycle myst be run multiple times during a single installation...

**Comment by smoors** [comment]
> > @boegel one solution would be to allow `buildopts` and friends be defined a list of strings, which get joined by EB with a space
> 
> `buildopts` can already be a list of string values, but that's used for indicating that the configure-build-test-install cycle myst be run multiple times during a single installation...

ah yes, I knew that..

**Comment by mboisson** [comment]
> this is the error for the intel version (foss builds fine). looks indeed like an issue with OSMesa linking, which was skipped previously.
> 
> ```
> [ 65%] Building CXX object CMakeFiles/vtkFiltersPointsCS.dir/CMakeFiles/vtkFiltersPointsCS/vtkLinearKernelClientServer.cxx.o
> /apps/brussel/CO7/skylake/software/impi/2019.7.217-iccifort-2020.1.217/intel64/bin/mpiicpc   -I/tmp/3254107.master01.hydra.brussel.vsc/tmp/vsc10009/build/ParaView/5.8.0/intel-2020a-Python-3.8.2-mpi/easybui
> ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaDestroyContext'
> ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaMakeCurrent'
> ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaGetCurrentContext'
> ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaCreateContext'
> ld: ../../../lib64/libvtkRenderingOpenGL2-pv5.8.so.5.8: undefined reference to `OSMesaGetProcAddress'
> make[2]: *** [bin/vtkProbeOpenGLVersion-pv5.8] Error 1
> make[2]: Leaving directory `/tmp/3254107.master01.hydra.brussel.vsc/tmp/vsc10009/build/ParaView/5.8.0/intel-2020a-Python-3.8.2-mpi/easybuild_obj'
> make[1]: *** [VTK/Rendering/OpenGL2/CMakeFiles/vtkProbeOpenGLVersion.dir/all] Error 2
> ```

actually, I get this error when building with GCC as well, and even if I use Gentoo's `libOSMesa.so`. 
Looking at the intermediate file, it looks like it does not link at all on `libOSMesa.so`. 

Looks like a ParaView bug to me.

**Comment by mboisson** [comment]
I created an issue on ParaView's gitlab... lets see what they say : 
https://gitlab.kitware.com/paraview/paraview/-/issues/20120

**Comment by mboisson** [comment]
Related posts : 
https://discourse.paraview.org/t/osmesa-linking-problems-compiling-paraview-on-opensuse-leap-15-1/4124/7
https://discourse.paraview.org/t/compiling-paraview-with-vtk-use-x-and-vtk-opengl-has-osmesa-simultaneously/4843

**Comment by Micket** [comment]
I'm here to complicate the situation a bit more: #10977 

So, I really really really want a version that works with VirtualGL, which will not work if linked with OSMesa.
I would also like this to be the version without any extra suffixes or such so that it's the primary dependency for OpenFOAM

Then, introduce secondary packages for
* ParaView-5.8.0-foss-2020a-Python-3.8.2-mpi-egl.eb
* ParaView-5.8.0-foss-2020a-Python-3.8.2-mpi-osmesa.eb

respectively for all your off-screen rendering needs. Probably these versions could just build the pvserver components since this is their usecase.

Since the ParaView we released in the last EB didn't actually build with OSMesa correctly, perhaps we can consider removing it from this PR, and then opening new PRs for a `osmesa`  and `egl` versions.

**Comment by mboisson** [comment]
I think these recipes actually address what you want to do : 
https://github.com/ComputeCanada/easybuild-easyconfigs/blob/computecanada-master/easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-gompi-2020a-offscreen-gpu.eb
https://github.com/ComputeCanada/easybuild-easyconfigs/blob/computecanada-master/easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-gompi-2020a-offscreen.eb
https://github.com/ComputeCanada/easybuild-easyconfigs/blob/computecanada-master/easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-gompi-2020a.eb

We build three flavors of ParaView. One which builds the client, with Qt and Mesa, one which builds fo offscreen rendering with OSMesa, and one which uses EGL from NVidia to work on GPUs.

File: easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-foss-2020a-Python-3.8.2-mpi.eb [hunk 1]
```diff
@@ -41,7 +41,7 @@ configopts += '-DPARAVIEW_ENABLE_FFMPEG=ON '
 
 # OpenGL & Mesa
 configopts += '-DOPENGL_glu_LIBRARY=$EBROOTLIBGLU/lib/libGLU.%s ' % SHLIB_EXT
-configopts += '-DOPENGL_INCLUDE_DIR=$EBROOTMESA/include'
+configopts += '-DOPENGL_INCLUDE_DIR=$EBROOTMESA/include '
 configopts += '-DVTK_OPENGL_HAS_OSMESA=ON '
```
Comment (Micket) [inline easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-foss-2020a-Python-3.8.2-mpi.eb:45]:
```suggestion
```

File: easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-intel-2020a-Python-3.8.2-mpi.eb [hunk 2]
```diff
@@ -46,7 +46,7 @@ configopts += '-DCMAKE_CXX_FLAGS_RELEASE="$CXXFLAGS -DNDEBUG -no-ipo" '
 
 # OpenGL & Mesa
 configopts += '-DOPENGL_glu_LIBRARY=$EBROOTLIBGLU/lib/libGLU.%s ' % SHLIB_EXT
-configopts += '-DOPENGL_INCLUDE_DIR=$EBROOTMESA/include'
+configopts += '-DOPENGL_INCLUDE_DIR=$EBROOTMESA/include '
 configopts += '-DVTK_OPENGL_HAS_OSMESA=ON '
```
Comment (Micket) [inline easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-intel-2020a-Python-3.8.2-mpi.eb:50]:
```suggestion
```

**Comment by boegel** [comment]
I'm happy with dropping `OSMESA=ON` from this Paraview easyconfig, but the implications are not 100% clear to me...

What does enabling `OSMESA` imply exactly (other than requiring `Mesa`)?

File: easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-foss-2020a-Python-3.8.2-mpi.eb (hunk 1 above)
Comment (boegel) [reply easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-foss-2020a-Python-3.8.2-mpi.eb:45]: Doesn't this imply dropping the `Mesa` dep too?

**Comment by Micket** [comment]
There are basically 3 modes, all, sort-of-mutually-exclusive, rendering backends:
1. Onscreen, classic GLX, like your classic glxgears
2. Offscreen option 1: OSMesa (off-screen-mesa). Not compatible with VirtualGL last i tried (and I can't imagine how it ever would  be). Software rendering only. This is a special non-glx part of Mesa.
3. Offscreen option 2: EGL (the glorious future). Supports hardware rendering off-screen. This is what you really want for pvserver (if you have GPUs). Mesa also supports EGL with software rendering, so, technically, this could superseed everything in the future. But, VirtualGL doesn't support EGL yet, only GLX. So, for Onscreen needs (via VNC, Xpra, ssh, with VirtualGL) you need GLX mode.

**Comment by boegel** [comment]
@Micket So that means picking between a `Paraview` that supports on screen rendering properly (with VirtualGL, no `OSMESA`) but no offscreen rendering, and one that supports offscreen rendering but no (proper) onscreen rendering?

Don't you basically need both onscreen and offscreen rendering when using Paraview as a dep for OpenFOAM?

Should the Paraview that has good support for on screen rendering be tagged with `-gui`?

**Comment by Micket** [comment]
@boegel Yes, it means picking. See ComputeCanada's easyconfigs above in @mboisson comments, where they offer all 3 flavours.

OpenFOAM build their own "parafoam" (instead of just writing a custom file format/file-reader plugin, they compile their own paraview :clown_face:  ) for post-processing.
OpenFOAM definitely isn't doing off-screen rendering for simulations if that was what you where worrying about, it's all post-processing, just like "normal" paraview usage, but with their hacked version to read their input format.
I have always removed the OSMesa stuff in all our ParaView builds locally because VirtualGL support is crucial 
ComputeCanada also uses the onscreen flavour as their dep for OpenFOAM.

Some of the versions need to be tagged, yes. My suggestion above was to rather tag the off-screen versions, like ComputeCanada does, though I'm not sure whether it's good to label it "offscreen-gpu" because, technically, EGL supports software rendering as well (though currently one needs to specify extra environment varialbes to coax it into doing so at runtime https://github.com/easybuilders/easybuild-easyconfigs/issues/10977#issuecomment-673634895 )

But, I don't really care what versionsuffixes. EGL support for VirtualGL is getting closer and closer, and with some simple patches to VTK, it should definitely be possible for EGL to superseed everything; onscreen and offscreen, gpu and software rendering, so, eventually we could just have 1 golder version again.

**Comment by boegel** [comment]
@Micket @mboisson I would like to get this merged ASAP...

What are the next steps here? The `Mesa` dependency needs to stay, right? So it's only a matter of test reports & merging?

**Comment by Micket** [comment]
No matter what we do, we need to have Mesa as a dependency. Mesa supplies the libraries for all 3 possible backends (good old GLX, the obscure "OSMESA" and the new-fangled EGL).

I suggest we
1. Drop OSMESA from this package, and consider this the on-screen version.
2. Introduce a new easyconfig with OSMESA support and add a suffix `-offscreen`
3. Introduce yet another easyconfig with EGL support and add a suffix `-offscreen-gpu`

(I have no particular strong feelings about what the suffix should be, just something along these lines)

**Comment by mboisson** [comment]
> No matter what we do, we need to have Mesa as a dependency. Mesa supplies the libraries for all 3 possible backends (good old GLX, the obscure "OSMESA" and the new-fangled EGL).
> 
> I suggest we
> 
> 1. Drop OSMESA from this package, and consider this the on-screen version.
> 2. Introduce a new easyconfig with OSMESA support and add a suffix `-offscreen`
> 3. Introduce yet another easyconfig with EGL support and add a suffix `-offscreen-gpu`
> 
> (I have no particular strong feelings about what the suffix should be, just something along these lines)

These would fit exactly what we do in Compute Canada, so I'm fine with those. For information, these are what we have installed: 
https://github.com/ComputeCanada/easybuild-easyconfigs/blob/computecanada-master/easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-gompi-2020a.eb
https://github.com/ComputeCanada/easybuild-easyconfigs/blob/computecanada-master/easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-gompi-2020a-offscreen.eb
https://github.com/ComputeCanada/easybuild-easyconfigs/blob/computecanada-master/easybuild/easyconfigs/p/ParaView/ParaView-5.8.0-gompi-2020a-offscreen-gpu.eb

**Comment by Micket** [comment]
If there aren't any objections, then I suggest we do this suggested change (which is what we were implicitly already doing here due to the typo) https://github.com/easybuilders/easybuild-easyconfigs/pull/10989#pullrequestreview-467070464 ?

**Comment by boegel** [comment]
> If there aren't any objections, then I suggest we do this suggested change (which is what we were implicitly already doing here due to the typo) [#10989 (review)](https://github.com/easybuilders/easybuild-easyconfigs/pull/10989#pullrequestreview-467070464) ?

Agreed.

@mboisson Can you also remove the `-DVTK_OPENGL_HAS_OSMESA=ON`?

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Review by Micket — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node3504.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/3dfb09d8d8a86ff91f3fa6994e194ce0 for a full test report.

**Comment by boegel** [comment]
Going in, thanks @mboisson!

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node2659.swalot.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/bb70488812dc41a8f4fa036b39b57ff7 for a full test report.

**Pull request merged**